src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A record mapper that is matching a GroupLayout to match the components of a record.
+ *
+ * @param <T> the Record type
+ */
+public final class LayoutRecordMapper<T extends Record>
+        implements Function<MemorySegment, T> {
+
+    enum Allow {EXACT, BOXING, BOXING_NARROWING_AND_WIDENING}
+
+    public static final Allow ALLOW = Allow.EXACT;
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+    private static final MethodHandles.Lookup PUBLIC_LOOKUP = MethodHandles.publicLookup();
+
+    private final Class<T> type;
+    private final GroupLayout layout;
+    private final MethodHandle ctor;
+
+    public LayoutRecordMapper(Class<T> type,
+                              GroupLayout layout) {
+        this(type, layout, 0);
+    }
+
+    private LayoutRecordMapper(Class<T> type,
+                               GroupLayout layout,
+                               long offset) {
+        this.type = type;
+        this.layout = layout;
+
+        var recordComponents = Arrays.asList(type.getRecordComponents());
+        var recordComponentNames = recordComponents.stream()
+                .map(RecordComponent::getName)
+                .collect(Collectors.toSet());
+
+        var nameToLayoutMap = layout.memberLayouts().stream()
+                // Only consider named layouts
+                .filter(l -> l.name().isPresent())
+                // Only look at the relevant layouts which could ever be matched with a record component
+                .filter(l -> recordComponentNames.contains(l.name().orElseThrow()))
+                .collect(toMap(l -> l.name().orElseThrow(), Function.identity(), throwingMerger()));
+
+        var missingComponents = recordComponents.stream()
+                .map(RecordComponent::getName)
+                .filter(l -> !nameToLayoutMap.containsKey(l))
+                .toList();
+
+        if (!missingComponents.isEmpty()) {
+            throw new IllegalArgumentException("There is no mapping for " +
+                    missingComponents + " in " + type.getName() +
+                    "(" + String.join(", ", recordComponents.stream().map(RecordComponent::getName).collect(Collectors.joining(", "))) + ")" +
+                    " provided by the layout " + layout);
+        }
+
+        var componentAndLayoutList = recordComponents.stream()
+                .map(c -> new ComponentAndLayout(c, nameToLayoutMap.get(c.getName())))
+                .toList();
+
+        Class<?>[] ctorParameterTypes = recordComponents.stream()
+                .map(RecordComponent::getType)
+                .toArray(Class<?>[]::new);
+
+        // An array of the record component MethodHandle extractors, each of type (MemorySegment)X
+        // where X is the component type.
+        MethodHandle[] handles = componentAndLayoutList.stream()
+                .map(cl -> {
```

Comment:
map() + filter(Object::isNull) should be a flatMap() or a mapMulti()

Reply:
The filter operation should be removed.